File: Backend/KaskoManagementSystem/Kasko.Business/Services/InsurancePackageService.cs
using Kasko.Business.DTOs.Package;
using Kasko.DataAccess.Repositories.Abstract;
using Kasko.Business.Interfaces;
using Kasko.Entities.Concrete;

namespace Kasko.Business.Services;

public class InsurancePackageService : IInsurancePackageService
{
    private readonly IUnitOfWork _unitOfWork;

    private readonly IGenericRepository<CoverageOption> _coverageOptions;

    public InsurancePackageService(
        IUnitOfWork unitOfWork,
        IGenericRepository<CoverageOption> coverageOptions)
    {
        _unitOfWork = unitOfWork;
        _coverageOptions = coverageOptions;
    }

    public async Task<IEnumerable<InsurancePackageDto>> GetAllAsync()
    {
        var packages = await _unitOfWork.InsurancePackages.GetAllAsync();

        var result = new List<InsurancePackageDto>();

        foreach (var package in packages.Where(x => x.IsActive && !x.IsDeleted))
        {
            var packageCoverages =
                await _unitOfWork.PackageCoverages.FindAsync(
                    x => x.InsurancePackageId == package.Id &&
                         !x.IsDeleted);

            var coverageDtos = new List<PackageCoverageDto>();

            foreach (var packageCoverage in packageCoverages)
            {
                var coverage =
                    await _unitOfWork.Coverages
                        .GetByIdAsync(packageCoverage.CoverageId);

                if (coverage == null ||
                    coverage.IsDeleted ||
                    !coverage.IsActive)
                {
                    continue;
                }

                coverageDtos.Add(
                    new PackageCoverageDto
                    {
                        CoverageId = coverage.Id,
                        CoverageName = coverage.Name,
                        CalculatedPrice = coverage.BasePrice,
                        IsDefault = packageCoverage.IsDefault,
                        Description = coverage.Description,
                        Options = (await _coverageOptions.FindAsync(
                                x => x.CoverageId == coverage.Id && !x.IsDeleted))
                            .OrderBy(x => x.SortOrder)
                            .Select(x => new CoverageOptionDto
                            {
                                Id = x.Id,
                                Name = x.Name,
                                Limit = x.Limit,
                                ExtraPrice = x.ExtraPrice,
                                IsDefault = x.IsDefault
                            })
                            .ToList()
                    });
            }

            result.Add(
                new InsurancePackageDto
                {
                    Id = package.Id,
                    Code = package.Code,
                    Name = package.Name,
                    Description = package.Description,
                    Factor = package.Factor,
                    IsActive = package.IsActive,
                    Coverages = coverageDtos
                });
        }

        return result;
    }
}

File: Fronted/KaskoManagementSystemFronted/src/app/features/quotes/insurance-package.service.ts
import { environment } from '../../../environments/environment';
import {
  Injectable,
  inject
} from '@angular/core';

import {
  HttpClient
} from '@angular/common/http';

import {
  Observable
} from 'rxjs';

export interface CoverageOption {
  id: string;
  name: string;
  limit: number | null;
  extraPrice: number;
  isDefault: boolean;
}

export interface PackageCoverage {
  coverageId: string;
  coverageName: string;
  calculatedPrice: number;
  isDefault: boolean;
  description?: string | null;
  options?: CoverageOption[];
}

export interface InsurancePackage {
  id: string;
  code: string;
  name: string;
  description?: string;
  factor: number;
  isActive: boolean;
  coverages: PackageCoverage[];
}

@Injectable({
  providedIn: 'root'
})
export class InsurancePackageService {

  private readonly http =
    inject(HttpClient);

  private readonly apiUrl =
    `${environment.apiBaseUrl}/InsurancePackage`;

  getPackages():
    Observable<InsurancePackage[]> {

    return this.http.get<InsurancePackage[]>(
      this.apiUrl
    );
  }
}